### PR TITLE
modern-cpp-Homework

### PR DESCRIPTION
**Pull request by Vicx95**
Exercises: 11, 14, 15, 16, 17, 18
Author: @Vicx95

### CODE REVIEW

**Review by MaciejSalamonski — APPROVED** [review]
Everything seems to be ok i think. Nice code 👍

**Review by jakjan95 — APPROVED** [review]
For me code looks good +1

File: shapes/main.cpp [hunk 1]
```diff
@@ -1,70 +1,75 @@
...
 
-void printAreas(const Collection& collection)
-{
-    for(vector<shared_ptr<Shape>>::const_iterator it = collection.begin(); it != collection.end(); ++it)
-        if(*it)
-            cout << (*it)->getArea() << std::endl;
+void printAreas(const Collection& collection) {
+    for (const auto& it : collection)
+        if (it)
+            cout << it->getArea() << std::endl;
```
Comment (WBigaj) [inline shapes/main.cpp]:
```suggestion
    for (const auto& it : collection) {
        if (it) {
            cout << it->getArea() << std::endl;
        }
    }
```

File: shapes/main.cpp [hunk 2]
```diff
@@ -1,70 +1,75 @@
...
 
-void printCollectionElements(const Collection& collection)
-{
-    for(Collection::const_iterator it = collection.begin(); it != collection.end(); ++it)
-        if(*it)
-            (*it)->print();
+void printCollectionElements(const Collection& collection) {
+    for (const auto& it : collection)
+        if (it)
+            it->print();
```
Comment (WBigaj) [inline shapes/main.cpp]:
```suggestion
    for (const auto& it : collection) {
        if (it) {
            it->print();
        }
    }
```

File: shapes/main.cpp [hunk 3]
```diff
@@ -1,70 +1,75 @@
...
-bool areaLessThan10(shared_ptr<Shape> s)
-{
-    if(s)
-        return (s->getArea() < 10);
+};
+auto areaLessThanX = [x{10}](shared_ptr<Shape> s) {
+    if (s) {
+        return (s->getArea() < x);
+    }
     return false;
```
Comment (WBigaj) [inline shapes/main.cpp]:
```suggestion
    return s ? (s->getArea() < x) : false;
```

File: shapes/main.cpp [hunk 4]
```diff
@@ -1,70 +1,75 @@
...
-}
+};
 
-bool perimeterBiggerThan20(shared_ptr<Shape> s)
-{
-    if(s)
+auto perimeterBiggerThan20 = [](shared_ptr<Shape> s) {
+    if (s)
         return (s->getPerimeter() > 20);
     return false;
```
Comment (WBigaj) [inline shapes/main.cpp]:
```suggestion
    return s ? (s->getPerimeter() > 20) : false;
```

File: shapes/main.cpp [hunk 5]
```diff
@@ -1,70 +1,75 @@
...
-typedef vector<shared_ptr<Shape>> Collection;
+using Collection = vector<shared_ptr<Shape>>;
 
-bool sortByArea(shared_ptr<Shape> first, shared_ptr<Shape> second)
-{
-    if(first == nullptr || second == nullptr)
+auto sortByArea = [](shared_ptr<Shape> first, shared_ptr<Shape> second) {
+    if (first == nullptr || second == nullptr)
         return false;
     return (first->getArea() < second->getArea());
```
Comment (WBigaj) [inline shapes/main.cpp]:
```suggestion
    if (first == nullptr || second == nullptr) {
        return false;
    }
    return (first->getArea() < second->getArea());
```

**Review by WBigaj — COMMENTED** [review]
I have checked your code and left my feedback.

**Review by ziobron — CHANGES_REQUESTED** [review]
Missing delegating constructors. This feature allows you to call other constructor from the same class on constructor's initalizer list.

**Review by ziobron — APPROVED** [review]

**Comment by ziobron** [comment]
15 XP za Nowoczesny C++ SFINAE (4447)
5 XP za Nowoczesny C++ alignas (4448)
5 XP za Nowoczesny C++ delegating constructors (4449)
5 XP za Nowoczesny C++ lambda (4450)
5 XP za Nowoczesny C++ structured bindings (4451)
15 XP za Nowoczesny C++ variadic templates (4452)

{ "requirement_ids": [4447, 4448, 4449, 4450, 4451, 4452] }

**Pull request closed without merging**